core/src/test/java/org/apache/calcite/test/JdbcTest.java

```diff
@@ -6908,6 +6910,51 @@ public void testJsonPretty() {
         .returns("A=29; B=35; C=37; D=36\n");
   }
 
+  /**
+   * Test case for
+   * <a href="https://issues.apache.org/jira/browse/CALCITE-3100">[CALCITE-3100]
+   * cast(? as DATE) won't work with PreparedStatement
+   * error</a>.
+   */
+  @Test public void testPreparedStatementException() throws Exception {
+    Connection connection = CalciteAssert.that()
+            .with(CalciteAssert.SchemaSpec.DATETIMEDATA).with(Lex.MYSQL).connect();
+
+    PreparedStatement dateStatement = connection.prepareStatement(
+            "select * from types where da = cast(? as DATE)");
+    dateStatement.setString(1, "1996-08-03");
+    dateStatement.execute();
+
+    PreparedStatement timeStatement = connection.prepareStatement(
+            "select * from types where ti = cast(? as TIME)");
+    timeStatement.setString(1, "10:15:00");
+    timeStatement.execute();
+
+    //From local time to UTC
+    SimpleDateFormat format = new SimpleDateFormat("HH:mm:ss", Locale.SIMPLIFIED_CHINESE);
+    java.util.Date d1 = format.parse("10:30:00");
+    java.util.Date d2 = format.parse("12:00:00");
+    TimeZone timeZone = TimeZone.getDefault();
+
+    long millD1 = d1.getTime() - timeZone.getRawOffset();
+    long millD2 = d2.getTime() - timeZone.getRawOffset();
+
+    String timeString1 = format.format(new java.util.Date(millD1));
+    String timeString2 = format.format(new java.util.Date(millD2));
+
```

Comment:
Can we just inline this `String timeString1 = format.format(new java.util.Date(millD1));` with `d1.getTime() - timeZone.getRawOffset();` BTW, can we reuse the tools in `DateTimeUtils` ?